feat: implement modules

bitwalker opened the pull request:
**NOTE:** I've broken up this PR into a sequence of commits that are smaller/more focused, and in the order that will give you the best context as you review each of them. These commits do not compile individually, but all together they do, as well as pass all of the parser crate tests - they are simply broken up this way to make it easier to review. **Please take note of the commit messages, they provide a lot of detail**.

---

This PR refactors the parser crate to implement modules, while in the process simplifying the grammar and the syntax tree itself. It adds support for rich diagnostics to all AST nodes, and makes heavy use of the new diagnostics infrastructure during program compilation/validation. It implements a generic visitor for implementing passes over the AST, and uses this both for import resolution and semantic analysis, both of which are also implemented in this PR. As an aside, this PR also removes the few vestiges of `match enf` currently in the compiler, which we can add back once the discussion around its syntax and semantics has been pinned down.

Now, this PR is a draft PR because after doing some initial review of how this affects upstream crates, namely the IR, I did not want to dive into updating the IR crate (which does not compile on this branch) until these changes have been signed off on, and we've discussed how to proceed. Many of the things which the IR crate does in terms of validation and constructing state for the constraint graph are actually made redundant by the changes in this PR, and I'm unsure how we want to approach that going forward.

To describe a bit what I mean: when you invoke the parser after these changes, you are asking the parser to produce a `Program`. In the process of producing that, the parser does all of the important work of semantic analysis, and in particular, resolving identifiers to their definitions, propagating types, etc. This means that when you get a `Program`, it is still a relatively high level representation, however all names have been resolved, most dead code has been eliminated, and it is ready for lowering to a form more amenable for optimization and execution (i.e. the IR). However, the IR (currently) is doing a lot of that work as it translates from the AST to the IR, which simply falls apart when modules are involved, and in general is not the ideal place in a compilation pipeline to do some of that stuff.

What I'd like to do to get the IR crate compiling again is first figure out how close the `Program` structure is to "ideal" form for lowering to the IR, and what we need to translate directly from it to the IR. It may be the case that some of the information gathered (and subsequently discarded) during semantic analysis, such as the dependency graph, would actually be useful during lowering to the IR, and if so, we might want to defer invoking semantic analysis as part of the parser, and instead put it in the hands of the IR lowering pass. If on the other hand, all of the information needed is already in the `Program` structure, then I suspect we can vastly simplify parts of the IR crate now.

grjte commented on parser/src/ast/expression.rs:
```diff
@@ -0,0 +1,719 @@
+        }
+    }
+
+    /// Returns true if the callee is a builtin function, e.g. `sum`
+    #[inline]
+    pub fn is_builtin(&self) -> bool {
+        self.callee.is_builtin()
+    }
```
This isn't allowed for evaluator function calls (maybe this is handled elsewhere, but we should at least note it here)

bitwalker replied:
I'm not sure I follow - you mean calling sum/prod from an evaluator?

grjte replied:
Ooh I think I misread this as caller instead of callee. Let's ignore this until I revisit :)

grjte commented on parser/src/ast/declarations.rs:
```diff
@@ -0,0 +1,523 @@
+
+/// Represents an item exported from a module
+///
+/// Currently, only constants and functions are exported.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum Export<'a> {
+    Constant(&'a crate::ast::Constant),
+    Evaluator(&'a EvaluatorFunction),
+}
```
We should be able to export pure functions as well (but maybe those just aren't here yet)

bitwalker replied:
Yep, that's just not added here because there is no `Function` type in the AST yet, I assumed we'd add that immediately after these modularity changes. The changes in this PR do anticipate functions though, and are prepped for when such a type is added.

grjte reviewed (commented):
This is a partial review. I'm not quite done going through the AST yet and haven't looked at tests/lexer/grammar changes, etc, so some of these comments/questions might be handled elsewhere, in which case just clarifying the comments a bit would help.

One question - so far it looks to me like we may have lost the functionality for applying selectors to constraints?

bitwalker commented on parser/src/ast/declarations.rs:
```diff
@@ -0,0 +1,523 @@
+///! There are no expressions/statements permitted in the top-level of a module, only declarations.
+///! These declarations define named items which are used by functions/constraints during evaluation.
+///!
+///! Some declarations introduce identifiers at global scope, i.e. they are implicitly defined in all
+///! modules regardless of imports. These are the `trace_columns`, `public_inputs`, and `random_values`
+///! sections.
```
Sure, that's simple to change - though it doesn't actually cause problems if those declarations are publically visible, and it definitely feels like a strange mix of visibilities from the root module, with only one of those key top-level declarations globally-visible while others aren't, especially because the `random_values` declaration doesn't have a fixed name. Library modules on their own are incomplete - they must be evaluated in the context of a root module to be meaningful, so referencing global declarations of any kind is likely to cause issues if they have different names.

That said, the fewer things globally visible the better.

grjte replied:
I agree that it's a strange mix of visibilities - I'd rather find an alternative way to deal with `random_values`, but we haven't come up with anything yet (we also haven't spent a lot of time on it).

bitwalker replied:
Probably the simplest would be to disallow naming the main random values declaration, i.e. changing the declaration syntax to either `random_values: [2]` or `random_values: [a, b]`; then making `$random` or `$rand` a predefined name to refer to the entire random_values array.

bitwalker commented on parser/src/ast/declarations.rs:
```diff
@@ -0,0 +1,523 @@
+/// The types of values which can be represented in an AirScript program
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum Type {
+    /// A scalar integer value
+    Integer,
```
Ah I did assume they were field elements, but hadn't seen them called that in the parser (mostly things like "number") - in any case this is an easy change to make with find/replace; how about `Felt`?

bitwalker commented on parser/src/ast/declarations.rs:
```diff
@@ -0,0 +1,523 @@
+    }
+}
+
+/// Declaration of a public input for an AIR. Public inputs are represented by a named identifier
+/// which is used to identify a fixed size array of length `size`.
+#[derive(Debug, Clone, Spanned)]
+pub struct PublicInput {
```
I'll read through the discussion to make sure nothing here gets in the way of that

bobbinth commented:
> It's represented in the AST this way because the syntax for such constraints and list comprehensions is virtually identical

I may be missing some context here - but I think the syntax is quite different:

```
# list comprehension
enf s[i]' = s[i + 1] for i in 2..15

# selectors
match enf:
  noop([stack_top])                     when op_flags[0]
  eqz([stack_top, stack_helpers[2]])    when op_flags[1]

# or selectors with new proposed syntax
match:
  when op_flags[0]: noop([stack_top])
  when op_flags[1]: eqz([stack_top, stack_helpers[2]])
```
Am I missing something?

bitwalker commented on parser/src/ast/statement.rs:
```diff
@@ -0,0 +1,85 @@
+impl Statement {
+    /// Checks this statement to see if it contains any constraints
+    ///
+    /// This is primarily necessary because `let` statements have a body, which is
+    /// also composed of statements, and so may be nested arbitrarily deep, containing
+    /// one or more constraints in its body.
```
This is used for lexical scoping - a constraint is only permitted in the _body_ of the `let`, not the value expression, i.e. the structure here is more like `let <name> = <value expression> in <body>` (without the `in`, as that isn't present in the original syntax). This further guarantees that there can never be a `let` followed by nothing, even if there were constraints expressed before it.

bitwalker commented on parser/src/ast/statement.rs:
```diff
@@ -0,0 +1,85 @@
+    }
+}
+
+/// A `let` statement binds `name` to the value of `expr` in `body`.
+#[derive(Clone, Spanned)]
+pub struct Let {
+    #[span]
+    pub span: SourceSpan,
+    /// The identifier to be bound
+    pub name: Identifier,
+    /// The expression to bind
+    pub value: Expr,
+    /// The scope in which the binding will be visible
+    pub body: Vec<Statement>,
+}
```
Imagine the syntax looking something like this for multiple lets:

```
let x = 1 + 1 in
    let y = x^2 in
       enf clk' = y
```

The body of `let x = 1 + 1` is `let y = ...`, and the body of `let y = x^2` is `enf clk' = y`. This provides natural lexical scoping during analysis/transformation, and ensures that a `let` always precedes at least one constraint.

grjte replied:
I see, thanks for the explanation

bitwalker replied:
I've fleshed out the docs for this to make it clear how the language syntax corresponds to the `Let` structure

bitwalker commented on parser/src/ast/expression.rs:
```diff
@@ -0,0 +1,719 @@
+pub enum ResolvableIdentifier {
+    /// This identifier is resolved to a local binding (i.e. function parameter or let-bound var)
+    Local(#[span] Identifier),
+    /// This identifier is resolved to a global binding (i.e. random values/public inputs/trace columns)
+    Global(#[span] Identifier),
```
Global here simply means any identifier visible globally, it has no additional semantics.

grjte replied:
My point is that the `trace_columns` and `public_inputs` are only visible in the root module where they're defined, whereas the `random_values` are visible everywhere

bitwalker replied:
Sure, I'll update the comment accordingly, but the notion of "global" still remains the same, even if that namespace is only notionally occupied by random values at this time.

grjte replied:
Yeah, my point was about the comment. If only random values are included then global is accurate

bitwalker commented on parser/src/ast/expression.rs:
```diff
@@ -0,0 +1,719 @@
+    IndexOutOfBounds,
+}
+
+/// [SymbolAccess] represents access to a named item in the source code; one of the following:
+///
+/// * A global name associated with trace columns, public inputs, or random values
+/// * A named constant
+/// * A module-local name associated with periodic columns
+/// * A let-bound variable
```
Those are also represented by this, yes, I'll make a note to add that to the doc comment here.

bitwalker commented:
@bobbinth The syntax doesn't change here, only the representation in the AST - but it's important to note that the `match` construct is not present in this PR. In my staged changes that bring it back, it has its own AST representation because it is considerably different syntactically than "comprehension" constraints.

bitwalker commented:
> I think I have a similar confusion to @bobbinth, that it doesn't feel natural to me to represent it as a list comprehension

Consider a portion of the current grammar (i.e. before this PR):

```
BoundaryConstraintExpr: BoundaryConstraint = {
    <column: SymbolAccess> "." <boundary: Boundary> "=" <value: BoundaryExpr>
        <comprehension: ConstraintComprehension<BoundaryExpr>?> =>

ConstraintComprehension<T>: ComprehensionContext = {
    <l:@L> "for" <members: Members> "in" <iterables: Iterables> <r:@R> =>? ...

ListComprehension<T>: ListComprehension = {
    <l:@L> <expr: T> "for" <members: Members> "in" <iterables: Iterables> <r:@R> =>? ...


```

The representation of these two things is identical, but the body of the comprehension is defined in the `BoundaryConstraint` rule for `ComprehensionConstraint`.

grjte commented:
> The representation of these two things is virtually identical, but the body of the comprehension is defined in the `BoundaryConstraint` rule for `ComprehensionConstraint`.

Yeah, I'm sure it will be clearer once I get to the new grammar. I guess we essentially treat

```
enf a = a + 1
```

as 

```
enf a = a + 1 for a in [a]
```

bitwalker commented on parser/src/ast/expression.rs:
```diff
@@ -0,0 +1,719 @@
+    BoundedSymbolAccess(BoundedSymbolAccess),
+    /// A binary operator over scalar values
+    Binary(BinaryExpr),
+    /// A call to a function or evaluator
+    Call(Call),
```
Sure thing, I'll flesh out the docs on the `Call` type to clarify why it has the type field at all, as well as why it is an `Option`

bitwalker replied:
The docs for `Call` have been updated, let me know if you think that helps clarify things!

grjte replied:
Thanks, I think the docs for `Call` are clear. I would still prefer to have a note here, though, similar to how you put one above `BoundedSymbolAccess`, because the docs for this say "Scalar expressions are expressions which evaluate to a single scalar value", but evaluator functions don't return a scalar value. I don't think it needs to be anything detailed since there are detailed docs elsewhere, but it would be good to have something for times when this is read independently of the `Call` docs.

bitwalker commented:
@grjte For plain constraints there is no need to wrap them in a comprehension, so `enf a = a + 1` remains that way. Specifically we're representing two variants of constraints in comprehension form:

* Actual constraint comprehensions, e.g. `enf a = a + 1 for a in [a] when k0` (this example has a selector, but that's optional of course)
* "Scalar" constraints with selectors, e.g. `enf a = a + 1 when k0`

The latter is parsed into comprehension form, as if it had been written as `enf a = a + 1 for i in 0..1 when k0`.

The result is that we end up with two types of constraints to deal with: simple, and comprehension. When `match` is added back, we'll have a third type which represents its unique semantics.

grjte reviewed (changes requested):
This is another partial review. It covers almost everything except the parser test files and the semantic analysis, which I'll finish tomorrow. I think the new design makes sense and the changes look great overall. Most of my comments are fairly small ones regarding clarity. At this point I would say I don't anticipate that I'll have any major change requests for this PR, so small issues aside it should be pretty much ready to merge once I get through the rest of it

bitwalker commented on parser/src/ast/trace.rs:
```diff
@@ -0,0 +1,311 @@
+    pub span: SourceSpan,
+    /// The index of this segment in the trace_columns declaration
+    pub id: TraceSegmentId,
+    /// The name of this trace segment, e.g. `$main`
+    ///
+    /// NOTE: The name of a trace segment is always a special identifier (i.e. has the `$` prefix)
```
Definitely agree that `main` and `aux` should not be special in any way, but I think there is probably value in making the top-level trace segment names global/special identifiers, i.e. if you name a segment `foo`, you should be able to reference a specific column of foo directly as `$foo[N]`. That said, I suspect in practice such a feature wouldn't be used much if at all, so perhaps it is only useful when performing certain transformations on the AST internally in the compiler.

grjte replied:
I guess my thinking was that a named trace segment `foo` is only accessible within the root module, whereas a named random values vector `rand` is actually globally accessible. Using `foo[i]` and `$rand[i]` distinguishes those things

bitwalker commented on parser/src/ast/trace.rs:
```diff
@@ -0,0 +1,311 @@
+impl ColumnBoundaryFlags {
+    /// An empty flag set that indicates the column is unconstrained
+    pub const EMPTY: Self = Self(0);
+    /// A flag set that indicates the column is constrained on the first boundary
+    pub const FIRST: Self = Self(0b001);
+    /// A flag set that indicates the column is constrained on the last boundary
+    pub const LAST: Self = Self(0b010);
+    /// A flag set that indicates the column is constrained on both boundaries
+    pub const BOTH: Self = Self(0b010);
```
Ah yes, not sure how I missed that, nice catch!

bitwalker commented on parser/src/ast/trace.rs:
```diff
@@ -0,0 +1,311 @@
+        self.boundary_constrained[column].is_constrained(boundary)
+    }
+
+    /// Marks `column` as constrained on `boundary`, and associates it with a span
+    /// that is responsible for the constraint.
+    ///
+    /// Returns `None` if the column was previously unconstrained on `boundary`,
+    /// otherwise it returns the span responsible for the previous constraint for
+    /// use in diagnostics
+    pub fn mark_constrained(
+        &mut self,
+        span: SourceSpan,
+        column: TraceColumnIndex,
+        boundary: Boundary,
+    ) -> Option<SourceSpan> {
+        let flags = &mut self.boundary_constrained[column];
+        if flags.is_constrained(boundary) {
+            Some(flags.span())
```
If it was already marked constrained on that boundary, the operation is a no-op (i.e. the flag is a single bit for a given boundary, so no action is required if already set); but if it was constrained on a different boundary, this conditional will not return true, and the second branch is taken which does set the constrained flag.

bitwalker commented on parser/src/ast/mod.rs:
```diff
@@ -1,125 +1,456 @@
+    pub evaluators: BTreeMap<QualifiedIdentifier, EvaluatorFunction>,
+    /// The set of used periodic columns referenced in this program.
+    pub periodic_columns: HashMap<QualifiedIdentifier, PeriodicColumn>,
+    /// The set of public inputs defined in the root module
+    ///
+    /// NOTE: Public inputs have globally-visible identifiers, so we do
+    /// not use [QualifiedIdentifier] as a key into this collection.
+    pub public_inputs: HashMap<Identifier, PublicInput>,
```
Ah, looks like I missed this comment when I updated things last week, thanks for catching that!

bitwalker commented on parser/src/lexer/mod.rs:
```diff
@@ -78,17 +78,25 @@ pub enum Token {
-    /// Used to declare AIR constraints module.
+    /// Used to declare AIR program.
     Def,
+    /// Used to declare AIR module.
+    Module,
```
In this particular instance I'm cool with either form, but I believe in our previous discussions, the topic of aggressive shorthand for things came up, and it sounded like the general consensus was to prefer more descriptive keywords unless there is a benefit to the shorthand. For example, `enf` is meaningless as a keyword (i.e. unless you already know that it is shorthand for "enforce"), but if you are writing a bunch of constraints, the shorthand is convenient, and is less noisy than the non-shorthand form, so it balances out. However, for less frequently used keywords, the shorthand is just unnecessarily concise (`ev` for example isn't even the typical shorthand for evaluator/evaluation, which would be "eval", so there is some cognitive overhead when encountered). For others, the shorthand is so common in the industry that there is no confusion about what they generally indicate (e.g. `def`, `fn`). In this instance I erred on the side of more descriptive, but `mod` is common enough in the industry that I don't have a particular problem with it vs `module`, so if we prefer `mod` here, that's an easy change to make. 

While I'm on the topic, `def` is a bit of an odd choice for module declarations, since it is typically used for function declarations in other languages where `def` or `define` is a keyword (e.g. Python, Elixir, Scheme/LISP, BASIC), but I think there is benefit in distinguishing between the root module and a library module, and `def` is already part of the language syntax, so changing it now is probably not worth it. I think I'd probably use `program`/`prog` for the root module, since it constitutes an AirScript "program", but ultimately it doesn't matter that much. I'm a programming language design/implementation nerd, so I probably care more about these things than is strictly necessary.

bobbinth replied:
My person preference (and not a strong one) is to go with `def/mod` - or if we don't like `def`, another option could be just `air` (I'm not sure about `prog` since this is not really a program - more like a set of polynomial equations).

bitwalker replied:
FYI: I've gone ahead and just updated everything to use `mod` for now

bitwalker commented on parser/src/parser/grammar.lalrpop:
```diff
@@ -57,399 +111,318 @@ Trace: Vec<Vec<TraceBinding>> = {
-        Ok(integrity_stmts)
+StatementBlock: Vec<Statement> = {
+    <Let> => vec![Statement::Let(<>)],
+    <stmts:ConstraintStatement+> <last:Let> => {
+        let mut stmts = stmts;
+        stmts.push(Statement::Let(last));
+        stmts
     },
```
For sure, always happy to answer questions to make review easier if it takes some of the load off :)

This case handles when a statement block contains a `let`, but starts with a non-let statement. `let` is always the last statement in its containing block because the "body" of the `let` are all of the statements in which the variable it binds is in scope. The other cases handle when the `let` comes first, and when the block has no `let` at all.

grjte replied:
Ahh yes, that makes sense, thanks!

bitwalker commented on parser/src/lexer/mod.rs:
```diff
@@ -78,17 +78,25 @@ pub enum Token {
+    /// Used to import an AIR module.
+    Use,
+    /// Used to import an item from an AIR module.
+    Import,
```
FYI: I've updated everything to use `use` only, with either `use module::item` or use module::*` for partial imports and wildcard imports respectively. I think we probably should consider adding `use module::{item1, item2}` at some point, but I haven't done that yet

grjte replied:
Sounds good to me. I agree with you about supporting `use module::{item1, item2}`, but it doesn't have to happen in this PR

bitwalker commented on parser/src/parser/grammar.lalrpop:
```diff
@@ -57,399 +111,318 @@ Trace: Vec<Vec<TraceBinding>> = {
+        } else {
+            // If a selector is provided for a scalar constraint, we convert it to a comprehension constraint to keep the AST simple
+            //
+            // The generated comprehension is equivalent to `enf <expr> for %0 in 0..1 when <selector>`
```
I've reworked this commentary to cover the entire `ConstraintExpr` rule, and to document more completely the different syntax variations which the grammar rule handles, the different constraint "types" which are represented in the AST, and how/why the different variations map to those constraint types. Since that rule is the only place where constraints are constructed by the parser, I think that should cover things pretty well now.

grjte commented on parser/src/ast/trace.rs:
```diff
@@ -0,0 +1,311 @@
+    }
+}
+
+/// [TraceAccess] is like [SymbolAccess], but is used to describe an access to a specific trace column or columns.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct TraceAccess {
+    /// The trace segment being accessed
+    pub segment: TraceSegmentId,
+    /// The index of the first column at which the access begins
+    pub column: TraceColumnIndex,
+    /// The number of columns being accessed
+    pub size: usize,
+    /// The offset from the current row.
+    ///
+    /// Defaults to 0, which indicates no offset/the current row.
+    ///
+    /// For example, if accessing a trace column with `a'`, where `a` is bound to a single column,
+    /// the row offset would be `1`, as the `'` modifier indicates the "next" row.
+    pub row_offset: usize,
+}
```
If we're only going to use it in the IR as a `Value` node in the graph, do you still prefer to define it here rather than just defining it in the IR?

bitwalker replied:
Just following up after our call, we'll move this to the IR crate

grjte commented on parser/src/parser/tests/mod.rs:
```diff
@@ -1,18 +1,619 @@
+        })
+    };
+
+    ($module:ident, $name:ident, $offset:literal, $ty:expr) => {
+        ScalarExpr::SymbolAccess(SymbolAccess {
+            span: miden_diagnostics::SourceSpan::UNKNOWN,
+            name: ResolvableIdentifier::Resolved(ident!($module, $name)),
+            access_type: AccessType::Default,
+            offset: $offset,
+            ty: Some($ty),
+        })
+    };
```
Do module and offset make sense in the same case, since offsets can only be applied to trace columns and those are always declared in the root module? (or is this related to handling of the evaluators? I haven't gone through all of the semantic analysis work yet)

bitwalker replied:
Yes this case can go away

grjte commented on parser/src/parser/tests/mod.rs:
```diff
@@ -1,18 +1,619 @@
+        })
+    };
+
+    ($module:ident, $name:ident [ $idx:literal ], $ty:expr) => {
+        ScalarExpr::SymbolAccess(SymbolAccess {
+            span: miden_diagnostics::SourceSpan::UNKNOWN,
+            name: ident!($module, $name).into(),
+            access_type: AccessType::Index($idx),
+            offset: 0,
+            ty: Some($ty),
+        })
+    };
```
Shouldn't we also have an `AccessType::Matrix` case for namespaced identifiers?

bitwalker replied:
Yep, it just didn't come up in any tests yet, so I hadn't added it

bitwalker replied:
I added this to make sure it's available when needed

grjte commented on parser/src/parser/tests/mod.rs:
```diff
@@ -1,18 +1,619 @@
+        ))
+    };
+
+    ($module:ident, $name:ident [ $idx:literal ], $offset:literal, $ty:expr) => {
+        ScalarExpr::SymbolAccess(SymbolAccess {
+            span: miden_diagnostics::SourceSpan::UNKNOWN,
+            name: ResolvableIdentifier::Resolved(ident!($module, $name)),
+            access_type: AccessType::Index($idx),
+            offset: $offset,
+            ty: Some($ty),
+        })
+    };
```
This seems like another case that shouldn't come up, since trace columns are only defined in the root module

grjte commented on parser/src/parser/tests/mod.rs:
```diff
@@ -1,18 +1,619 @@
+    };
+}
+
+macro_rules! band {
+    ($lhs:expr, $rhs:expr) => {
+        mul!($lhs, $rhs)
+    };
+}
+
+macro_rules! bor {
+    ($lhs:expr, $rhs:expr) => {{
+        sub!(add!($lhs, $rhs), mul!($lhs, $rhs))
+    }};
+}
```
I'd prefer `and` and `or`, and since we only have one type I don't think we risk lack of clarity

grjte commented on parser/src/parser/tests/mod.rs:
```diff
@@ -1,18 +1,619 @@
+                name: ResolvableIdentifier::Local(ident!($name)),
+                access_type: AccessType::Default,
+                offset: 0,
+                ty: Some($ty),
```
We should never be able to have a non-Felt type for a bounded symbol access (although maybe this is for testing invalid cases?)

bitwalker replied:
Yeah that can probably be hardcoded here. If it ever comes up that we need to specify arbitrary types for certain tests, we can handle that as an additional case.

bitwalker replied:
Ah, I know why this exists at the moment - for tests which do not run through semantic analysis, the AST nodes are untyped, so we need a case that is untyped, and one that is typed, even if that type is always the same. So I made `bounded_access` mirror `access` here.

grjte replied:
Makes sense. It might be clearer for the future to add a comment which says that

grjte commented on parser/src/parser/tests/mod.rs:
```diff
@@ -1,18 +1,619 @@
+        ))
+    };
+
+    ($module:ident, $name:ident, $bound:expr, $ty:expr) => {
+        ScalarExpr::BoundedSymbolAccess(BoundedSymbolAccess::new(
+            miden_diagnostics::SourceSpan::UNKNOWN,
+            SymbolAccess {
+                span: miden_diagnostics::SourceSpan::UNKNOWN,
+                name: ResolvableIdentifier::Resolved(ident!($module, $name)),
+                access_type: AccessType::Default,
+                offset: 0,
+                ty: Some($ty),
+            },
+            $bound,
+        ))
+    };
```
similar question/comment to previously about only trace columns only being defined in the root module. There's another case below as well

bitwalker replied:
Yep, this can go away. I created a lot of these upfront before I knew which ones I'd actually need - I was anticipating that I might need to create invalid structures in some cases, but that didn't really happen, so a number of these macros have cases that can be eliminated.

grjte commented on parser/src/parser/tests/mod.rs:
```diff
@@ -1,18 +1,619 @@
+        ListComprehension::new(miden_diagnostics::SourceSpan::UNKNOWN, $body, context, None)
+    }};
+
+    (($(($binding:ident, $iterable:expr)),*) => $body:expr, when $selector:expr) => {{
+        let context = vec![
+            $(
+                (ident!($binding), $iterable)
+            ),+
+        ];
+        ListComprehension::new(miden_diagnostics::SourceSpan::UNKNOWN, $body, context, Some($selector))
+    }};
+
+    (($(($binding:literal, $iterable:expr)),*) => $body:expr, when $selector:expr) => {{
```
this is such a tiny nit, but I would probably drop the comma before `when`. It doesn't mirror the language, so I don't feel like it's helpful, and I think it would make me likely to make an error when using the macro. It's not a big deal though

bitwalker replied:
I absolutely would have done that (and in fact attempted to initially), but the `expr` pattern type can only be followed by a handful of tokens, and `when` is not one of those. I decided to separate the selector into a separate argument, but kept the `when` thinking it would be easier to read the macro with it present than with only a comma.

Procedural macros could work magic here, but that has a much higher implementation cost which wasn't worth it in my opinion

grjte replied:
Yep, makes sense. I wondered if that might be the case

grjte reviewed (commented):
This is another partial review which now mostly covers the parser tests. (Semantic analysis is still remaining)

One broad comment regarding the parser test macros - some of these are quite complex and I would feel more comfortable if we had a couple of tests to cover those ones. Adding some docs & comments would help as well for the macros with many cases

bitwalker commented on parser/src/ast/mod.rs:
```diff
@@ -1,125 +1,458 @@
+        while let Some(module_name) = visitor.next(&modgraph) {
+            // Remove the module from the library temporarily, so that we
+            // can look up other modules in the library while we modify it
+            let mut module = library.modules.remove(&module_name).unwrap();
```
I generally use `unwrap` when `unreachable!()` would apply - i.e. it should never be possible for the unwrap to fail. In this specific case, we know the `library` has all of the modules in the dependency graph, or we would have already raised an error in semantic analysis. Likewise in the other comment you left to the same effect - all of the dependencies are known to exist or we would have already bailed with an error. On reflection though, I think I'll update these to use the indexing syntax, it's cleaner/clearer and doesn't require us to debate the value of `unwrap` vs `expect` in these cases, and IIRC uses `expect` internally anyway.

bitwalker commented on parser/src/parser/tests/variables.rs:
```diff
@@ -1,179 +1,130 @@
+    let mut expected = Module::new(ModuleType::Library, SourceSpan::UNKNOWN, ident!(test));
+    // The constraint is converted into a comprehension constraint by the parser, which
+    // involves generating an iterable with one element and giving it a generated binding
+    let body = vec![let_!(flag = expr!(band!(access!(n1), not!(access!(n2)))) =>
+                  enforce_all!(lc!((("%1", range!(0..1))) => eq!(access!(clk, 1), add!(access!(clk), int!(1))), when access!(flag))))];
```
Here's what that code would look like macro-expanded:

```rust
Statement::Let(Let::new(
    SourceSpan::UNKNOWN, 
    ident!(flag), 
    expr!(band!(access!(n1), not!(access!(n2)))),
    enforce_all!(ListComprehension::new(
        SourceSpan::UNKNOWN, 
        eq!(access!(clk, 1), add!(access!(clk), int!(1))), 
        vec![(ident!("%1"), range!(0..1))], 
        Some(access!(flag))
    ))
)
```

It isn't too bad in this case, as there is only one `let` and one comprehension; and the comprehension only binds a single iterable which is a constant, but once you get into nested `let` statements, or complex comprehensions, it gets noisy real quick.

Overall, the benefit of the macros is three-fold - making it easy to write assertions against the AST by using macros that approximate the syntax (to the degree possible using macros-by-example); ensuring that changes to the AST do not require updating hundreds of tests (instead only the macro definition need be updated); and reducing the amount of unnecessary noise that instantiating the AST structs typically involves, putting the focus on the syntax structure being defined.

The `let_` and `lc` macros in particular are effective at reducing a lot of noise, as they are the most complicated structures in the AST. That said, it's all tradeoffs - it would obviously be preferable to be able to match the actual language syntax better, as macros-by-example have some odd quirks and limitations, and there isn't a great way to format deeply nested macros, so it can be a bit dense when reading them. On the other hand, the macros here make the tests visible on a single page of code, which means you can see the code being parsed and the macros corresponding to the expected AST at the same time, which is the biggest benefit in my mind, and go a long way to making the macro mess more obvious/clear. I'm open to removing/changing macros which are more of a problem than a help though.

grjte replied:
Yep, I can imagine how messy it will get with blocks of multiple let expressions, even e.g. just an evaluator with 4-5 `let` and constraint expressions. I'm ok with keeping the macros, but we should see what @bobbinth thinks

bitwalker commented on parser/src/parser/tests/mod.rs:
```diff
@@ -1,18 +1,619 @@
+    };
+}
+
+#[allow(unused)]
+macro_rules! global {
+    ($name:ident, $offset:literal, $ty:expr) => {
+        ScalarExpr::SymbolAccess(SymbolAccess {
+            span: miden_diagnostics::SourceSpan::UNKNOWN,
+            name: ResolvableIdentifier::Global(ident!($name)),
+            access_type: AccessType::Default,
+            offset: $offset,
+            ty: Some($ty),
+        })
+    };
+
+    ($name:ident, $ty:expr) => {
+        ScalarExpr::SymbolAccess(SymbolAccess {
+            span: miden_diagnostics::SourceSpan::UNKNOWN,
+            name: ResolvableIdentifier::Global(ident!($name)),
+            access_type: AccessType::Default,
+            offset: 0,
+            ty: Some($ty),
+        })
+    };
+
+    ($name:literal, $ty:expr) => {
+        ScalarExpr::SymbolAccess(SymbolAccess {
+            span: miden_diagnostics::SourceSpan::UNKNOWN,
+            name: ResolvableIdentifier::Global(ident!($name)),
+            access_type: AccessType::Default,
+            offset: 0,
+            ty: Some($ty),
+        })
+    };
```
Actually I think the first case is still useful as well, since you can have an expression like `$rand[0]`, which the first case is designed to handle

grjte replied:
Should we remove the first case then?

grjte commented on parser/src/parser/tests/integrity_constraints.rs:
```diff
@@ -361,298 +360,232 @@ fn integrity_constraint_with_indexed_trace_access() {
+    ));
+    expected.integrity_constraints = Some(Span::new(
+        SourceSpan::UNKNOWN,
+        vec![enforce_all!(
+            lc!(((x, range!(1..4))) => eq!(access!(x), add!(access!(a), access!(b))))
+        )],
```
note: this is another constraint that doesn't make sense. `enf 1 = a + b` would be fine, but if we enforce that `a+b` is 4 different things then only one of the constraints will hold, so constraint evaluation will fail.

grjte replied:
Yep, I know almost all of these errors were there before, but I think they're actually problematic now, so I figured pointing them all out at this point is the best way to identify additional changes that need handling. Since they're all minor changes I probably prefer to adjust them now since they're all identified in this review, but if you have a strong preference to do it in a follow-up PR I think that's ok.

grjte commented on parser/src/parser/tests/list_comprehension.rs:
```diff
@@ -1,437 +1,234 @@
+    expected.boundary_constraints = Some(Span::new(
+        SourceSpan::UNKNOWN,
+        vec![
+            let_!(x = lc!(((col, expr!(access!(c)))) => exp!(access!(col), int!(7))).into() =>
+                  enforce!(eq!(bounded_access!(a, Boundary::First), add!(add!(add!(access!(x[0]), access!(x[1])), access!(x[2])), access!(x[3]))))),
```
note: As mentioned previously, this shouldn't really be valid, since we can't access the "current" or "next" value of columns in boundary constraints. An integrity constraint formed this way would be valid (assuming it's enforced against `a` rather than `a.first`). Changing the list comprehension to something else that doesn't reference a trace column would make it valid, e.g. `let x = [i^7 for i in 0..4]`.

grjte commented on parser/src/parser/tests/list_comprehension.rs:
```diff
@@ -1,437 +1,234 @@
+    expected.boundary_constraints = Some(Span::new(
+        SourceSpan::UNKNOWN,
+        vec![
+            let_!(x = lc!(((i, range!(0..3)), (c, expr!(access!(c)))) => mul!(exp!(int!(2), access!(i)), access!(c))).into() =>
+                  enforce!(eq!(bounded_access!(a, Boundary::First), add!(add!(add!(access!(x[0]), access!(x[1])), access!(x[2])), access!(x[3]))))),
```
note: similar issue to the above here - this is invalid, but could be made valid by removing the trace column from the list comprehension. The boundary constraint tests below all seem to have the same issue, but I won't note it again

grjte commented on parser/src/parser/tests/list_comprehension.rs:
```diff
@@ -440,455 +237,193 @@ fn ic_one_iterable_identifier_lc() {
+    expected.integrity_constraints = Some(Span::new(
+        SourceSpan::UNKNOWN,
+        vec![
+            let_!(x = lc!(((c, expr!(slice!(c, 0..3)))) => access!(c)).into() =>
+                   enforce!(eq!(access!(a), add!(add!(add!(access!(x[0]), access!(x[1])), access!(x[2])), access!(x[3]))))),
```
note: this shouldn't be valid, because the range syntax should not be inclusive, so x only has length 3

grjte commented on parser/src/parser/tests/list_comprehension.rs:
```diff
@@ -440,455 +237,193 @@ fn ic_one_iterable_identifier_lc() {
+        SourceSpan::UNKNOWN,
+        vec![
+            let_!(diff = lc!(((x, expr!(access!(c))), (y, expr!(access!(d)))) => sub!(access!(x), access!(y))).into() =>
+                  enforce!(eq!(access!(a), add!(add!(add!(access!(x[0]), access!(x[1])), access!(x[2])), access!(x[3]))))),
```
note: `x` is not defined in this context, which was probably not intended for this test. It was probably supposed to be `diff`

grjte commented on parser/src/parser/tests/constant_propagation.rs:
```diff
@@ -0,0 +1,117 @@
+    const B = [[1, 1], [2, 2]]
+
+    integrity_constraints:
+        enf test_constraint(b)
```
this should be `enf test_constraint([b])` here and in the comment below. It shouldn't be valid to call an evaluator without the brackets that indicate the trace segment(s)

grjte reviewed (commented):
This review completes all of the test files. There are a bunch of tests with invalid AirScript that was previously acceptable at the parsing stage and addressed in the IR (or in some cases not addressed yet, but with issues open). I've marked everything that is invalid or nonsensical, since most validation is now moving to the semantic analysis pass.

It probably makes sense to adjust the tests to valid AirScript during this PR, but the checks for each of these things could be part of the next PR(s) if you think that makes more sense.

grjte commented on parser/src/transforms/constant_propagation.rs:
```diff
@@ -0,0 +1,486 @@
+                    Err(err) => ControlFlow::Break(err),
+                }
+            }
+            // While calls cannot be constant folded, arguments can be
+            ScalarExpr::Call(ref mut call) => self.visit_mut_call(call),
```
note: but only for pure functions

bitwalker replied:
Yes, though it's sort of a meaningless distinction here, since it would never be possible to constant fold arguments to an evaluator anyway (the arguments can't ever be constant). Semantic analysis should catch any invalid evaluator arguments so that we don't have to consider that here.

grjte replied:
Yeah, I know it will just do nothing (assuming things are working correctly). I guess I just like precision in comments because I feel like it helps people who are working with the code & saves time. Things that are written in comments don't have to be known or considered. It's a tiny nit, so you can ignore it if you disagree

grjte commented on parser/src/transforms/constant_propagation.rs:
```diff
@@ -0,0 +1,486 @@
+            //
+            // We deal with symbol accesses directly, as they may evaluate to an aggregate constant
+            Expr::SymbolAccess(ref mut access) => {
+                let constant_value = match access.name {
+                    // Possibly a reference to a constant declaration
+                    ResolvableIdentifier::Resolved(ref qid) => {
+                        self.global.get(qid).cloned().map(|s| (s.span(), s.item))
+                    }
+                    // Possibly a reference to a local bound to a constant
+                    ResolvableIdentifier::Local(ref id) => {
+                        self.local.get(id).cloned().map(|s| (s.span(), s.item))
+                    }
+                    // Other identifiers cannot possibly be constant
+                    _ => None,
+                };
```
this code is duplicated in the function above, so I'd prefer to use a helper in both cases

grjte reviewed (commented):
I've now gone through all files, although I haven't gone through the semantic analysis file in full depth, since I'm expecting it to change somewhat as a result of this review and possibly some of the comments on the tests

bitwalker commented on parser/src/transforms/constant_propagation.rs:
```diff
@@ -0,0 +1,486 @@
+                    Ok(exp) => l.item.pow(exp),
+                    Err(_) => return Err(InvalidConstantError::InvalidExponent(expr.span())),
+                },
+                // This op cannot be folded
+                BinaryOp::Eq => return Ok(None),
```
Good point, I was assuming we'd catch this during semantic analysis, but we can also raise validation errors at later stages such as constant propagation as we do more in-depth analysis/transformation. I'll make a note to do so in one place or the other

bitwalker commented on parser/src/sema/semantic_analysis.rs:
```diff
@@ -0,0 +1,1607 @@
+    TraceParam(TraceRef),
+    /// A direct reference to one or more contiguous trace columns
+    TraceColumn(TraceRef),
+    /// A potentially non-contiguous set of trace columns
+    Vector(Vec<BindingType>),
```
It can be yes, though this type has changed a bit in the PR I'm preparing for the lowering to IR

grjte replied:
Ok, I'll look at it there

bitwalker commented on parser/src/sema/semantic_analysis.rs:
```diff
@@ -0,0 +1,1607 @@
+    }
+
+    // Visit scalar constraints and ensure that they are valid semantically, and have correct types
+    fn visit_mut_enforce(&mut self, expr: &mut ScalarExpr) -> ControlFlow<ModuleError> {
```
Yeah, it's been on my TODO list to clean up this module a bit. I do prefer to keep the `SemanticAnalysis` implementation in a single file, as it makes navigating it much easier, but some of the auxilary types and functions can be moved out, and even within the implementation some of the functions definitely need to be broken up/factored out. I've already done some cleanup in my PR for the IR lowering, so it might make more sense to handle the clean up there since they need to be merged at the same time anyway, but if you think it would make review easier I can do some of the clean up here first and rebase on top of it.

grjte replied:
Hmm no, let's leave it as it is if you've already done some of that. I'll just defer a more thorough review of the semantic analysis to the next PR

bitwalker commented on parser/src/parser/tests/variables.rs:
```diff
@@ -1,179 +1,130 @@
+    // The constraint is converted into a comprehension constraint by the parser, which
+    // involves generating an iterable with one element and giving it a generated binding
+    let body = vec![let_!(flag = expr!(and!(access!(n1), not!(access!(n2)))) =>
+                  enforce_all!(lc!((("%1", range!(0..1))) => eq!(access!(clk, 1), add!(access!(clk), int!(1))), when access!(flag))))];
```
A generated variable, it is specifically made inexpressible in the syntax to guarantee generated variables can't conflict with any user-expressed identifier.

bitwalker commented on parser/src/parser/tests/sections.rs:
```diff
@@ -6,6 +6,10 @@ use super::ParseTest;
 fn error_constraint_without_section() {
     // Constraints outside of valid sections are not allowed.
-    let source = "enf clk' = clk + 1";
+    let source = r#"
+    def test
+
+    enf clk' = clk + 1
+    "#;
```
Yeah I was thinking the same, seems like it probably predates some of the other test modules that deal with sections

grjte reviewed (approved):
Thanks a lot for all this work @bitwalker. Things look great overall. There are still a few open comments, but I'll be away for the next few days and I don't want to block this over them, so I'm going to approve at this point. 

To summarize the remaining open items:

- tests: we should just decide if we'll make these valid AirScript now (my preference) or in a future PR (also ok). We should also make a note of new checks that need to be added to the semantic analysis based on the things I noted in those test files
- semantic analysis: I haven't done a very thorough review here & would like to see a minor refactor to make functions a bit smaller & easier to read, but this can happen in a future PR, since you've already started doing that there.
- constant propagation file: I requested a few minor refactors here. This can also be deferred to a future PR if you prefer.

bitwalker commented:
@hackaugusto I've updated the `air-codegen-masm` crate as part of rebasing this PR on `next` before I merge it, and there are 5 broken tests which I think are just due to changes in what IR is produced from a given source program + optimizations (e.g. constant prop), but would like to verify.

The commit which modifies the MASM code generator is b1440c1, and the broken tests are:

* `codegen::masm::bitwise`
* `codegen::masm::constants`
* `codegen::masm::list_comprehension`
* `codegen::masm::pub_inputs`
* `codegen::masm::variables`

I suspect we can just update the expected output to account for the changes to the IR lowering phase of the compiler, but don't want to do that without checking with you first. Let me know if you see any bugs.

@bobbinth @grjte This branch is now rebased against the latest code in `next`, so this is ready to merge once @hackaugusto  signs off on updating those tests.

hackaugusto commented:
Yeah, it is because of the new optimizations, it seems to be constants propagation:

- subtraction by zero are gone
- exponentiation of a constant by a constant are pre-computed
- basic operations like add/sub/mul of constants is also pre-computed

and there seems to be a case of https://github.com/0xPolygonMiden/air-script/issues/314 in there.

Thank you for handling the rebase!

bitwalker commented:
Thanks @hackaugusto! You're right that this changeset includes a patch that orders declarations by name in the intermediate representation, which was done for determinism purposes, but seems to also solve #314, so we may be able to close that out when this is merged
